storage/internal/elastic/store.go

```diff
@@ -0,0 +1,370 @@
+	}
+	defer res.Body.Close()
+
+	if res.IsError() {
+		var e map[string]interface{}
+		if err := json.NewDecoder(res.Body).Decode(&e); err != nil {
+			log.Printf("error parsing the response body: %s", err)
+		} else {
+			// Print the response status and error information.
+			log.Printf("[%s] %s: %s",
+				res.Status(),
+				e["error"].(map[string]interface{})["type"],
+				e["error"].(map[string]interface{})["reason"],
+			)
+		}
+	}
+
+	var r lastIndexResponse
+	if err := json.NewDecoder(res.Body).Decode(&r); err != nil {
+		log.Printf("Error parsing the response body: %s", err)
+		return 0, err
+	}
+
+	hits := r.hits.total
+	var lastIndex uint64 = 0
+	if hits > 0 {
+		lastIndex, err = parseInt64StringToUint64(r.aggregations.logsOrEvents.lastIndex.value)
+		if err != nil {
+			return lastIndex, err
+		}
+	}
+
+	// Print the response status, number of results, and request duration.
+	log.Debugf(
+		"[%s] %d hits; lastIndex: %d",
+		res.Status(),
+		hits,
+		lastIndex,
+	)
+
+	return lastIndex, nil
+
+}
+
+// List simulates long polling request by :
+// - periodically querying ES for documents (Aggregation to get the max iid and 0 size result).
+// - if a some result is found, wait some time (es_refresh_wait_timeout) in order to:
+//   	- let ES index recently added documents AND to let
+// 		- let Yorc eventually Set a document that has a less iid than the older known document in ES (concurrence issues)
+// - if no result if found after the the given 'timeout', return empty slice
+func (s *elasticStore) List(ctx context.Context, k string, waitIndex uint64, timeout time.Duration) ([]store.KeyValueOut, uint64, error) {
+	log.Debugf("List called k: %s, waitIndex: %d, timeout: %v", k, waitIndex, timeout)
+	if err := utils.CheckKey(k); err != nil {
+		return nil, 0, err
+	}
+
+	// Extract indice name by parsing the key
+	storeType, deploymentID := extractStoreTypeAndDeploymentID(k)
+	indexName := getIndexName(s.cfg, storeType)
+	log.Debugf("storeType is: %s, indexName is: %s, deploymentID is: %s", storeType, indexName, deploymentID)
+
+	query := getListQuery(s.clusterID, deploymentID, waitIndex, 0)
+
+	now := time.Now()
+	end := now.Add(timeout - s.cfg.esRefreshWaitTimeout)
+	log.Debugf("Now is : %v, date after timeout will be %v (ES timeout duration will be %v)", now, end, timeout-s.cfg.esRefreshWaitTimeout)
+	var values = make([]store.KeyValueOut, 0)
+	var lastIndex = waitIndex
+	var hits = 0
+	var err error
+	for {
+		// first just query to know if they is something to fetch, we just want the max iid (so order desc, size 1)
+		hits, values, lastIndex, err = doQueryEs(s.esClient, indexName, query, waitIndex, 1, "desc")
+		if err != nil {
+			return values, waitIndex, errors.Wrapf(err, "Failed to request ES logs or events, error was: %+v", err)
+		}
+		now := time.Now()
+		if hits > 0 || now.After(end) {
+			break
+		}
+		log.Debugf("hits is %d and timeout not reached, sleeping %v ...", hits, s.cfg.esQueryPeriod)
+		time.Sleep(s.cfg.esQueryPeriod)
```

Comment:
You could use a ticker here instead of a time.sleep as done here: https://github.com/ystia/yorc/blob/develop/storage/internal/file/store.go#L446